packages/crypto-wallet-core/src/transactions/xrp/index.ts

```diff
@@ -1,64 +1,109 @@
+import { Instructions, TransactionJSON } from 'ripple-lib/dist/npm/transaction/types';
 import { Key } from '../../derivation';
 
+enum HashPrefix {
+  // transaction plus signature to give transaction ID
+  TRANSACTION_ID = 0x54584E00
+}
 export class XRPTxProvider {
   create(params: {
     recipients: Array<{ address: string; amount: string }>;
-    data: string;
-    tag: number;
-    sourceAddress: string;
-    invoiceID: string;
-    fee: string;
+    tag?: number;
+    from: string;
+    invoiceID?: string;
+    fee: number;
     nonce: number;
   }) {
-    const { recipients, tag, sourceAddress, invoiceID, fee, nonce } = params;
+    const { recipients, tag, from, invoiceID, fee, nonce } = params;
     const { address, amount } = recipients[0];
+    const rippleAPI = new RippleAPI();
+    const { schemaValidate } = RippleAPI._PRIVATE.schemaValidator;
+    const Flags = 2147483648;
+    const amountStr = rippleAPI.dropsToXrp(amount);
+    const feeStr = rippleAPI.dropsToXrp(fee.toString());
     const payment: Payment = {
       source: {
-        address: sourceAddress,
-        tag: tag || undefined,
+        address: from,
         maxAmount: {
-          value: amount.toString(),
+          value: amountStr,
           currency: 'XRP'
         }
       },
       destination: {
         address,
-        tag: tag || undefined,
         amount: {
-          value: amount.toString(),
+          value: amountStr,
           currency: 'XRP'
         }
-      },
-      invoiceID: invoiceID || undefined,
+      }
     };
-
     const instructions: Instructions = {
-      fee,
+      fee: feeStr,
       sequence: nonce,
       maxLedgerVersion: null,
     };
-
-    let rippleAPI = new RippleAPI();
-    return rippleAPI.preparePayment(sourceAddress, payment, instructions).then((preparedTx) => {
-      return preparedTx.txJSON;
-    }).catch((err) => {
-      return;
-    });
+    const txJSON: TransactionJSON = {
+      TransactionType: 'Payment',
+      Account: from,
+      Destination: address,
+      Amount: amount.toString(),
+      Flags,
+      Fee: fee.toString(),
+      Sequence: nonce
+    };
+    if (invoiceID) {
+      payment.invoiceID = invoiceID;
+      txJSON.InvoiceID = invoiceID;
+    }
+    if (tag) {
+      payment.destination.tag = tag;
+      txJSON.DestinationTag = tag;
+    }
+    schemaValidate('preparePaymentParameters', { address: from, payment, instructions});
+    schemaValidate('tx-json', txJSON);
+    return rippleBinaryCodec.encode(txJSON);
   }
 
-  sign(params: { tx: string; key: Key; }) {
+  getSignatureObject (params: { tx: string; key: Key; }) {
     const { tx, key } = params;
-    const txJSON = tx;
+    const txJSON = rippleBinaryCodec.decode(tx);
     let rippleAPI = new RippleAPI();
-    const signedTx = rippleAPI.sign(txJSON, {
-      privateKey: key.privKey,
-      publicKey: key.pubKey,
+    const signedTx = rippleAPI.sign(JSON.stringify(txJSON), {
+      privateKey: key.privKey.toUpperCase(),
```

Comment:
why does this need to be uppercase?

Reply:
Ripple validation throws an error if not uppercase.

Reply:
```
1) Transaction Sign
       should not fail to sign a XRP tx with lowercase key:
     Error: `e8` is not a valid hex representation of a byte
      at byteForHex (node_modules/ripple-binary-codec/distrib/npm/utils/bytes-utils.js:38:11)
      at parseBytes (node_modules/ripple-binary-codec/distrib/npm/utils/bytes-utils.js:52:18)
      at new Blob (node_modules/ripple-binary-codec/distrib/npm/types/blob.js:9:21)
      at Function.from (node_modules/ripple-binary-codec/distrib/npm/types/blob.js:22:14)
      at /Users/justin/bitpay/bitcore/packages/crypto-wallet-core/node_modules/ripple-binary-codec/distrib/npm/types/st-object.js:31:46
```

Reply:
If I remove `.toUpperCase()` this error returns.

Reply:
Strange!